op-service/sources/receipts.go

```diff
@@ -4,18 +4,212 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/ethereum/go-ethereum/trie"
 
 	"github.com/ethereum-optimism/optimism/op-service/eth"
 )
 
+type ReceiptsProvider interface {
+	// FetchReceipts returns a block info and all of the receipts associated with transactions in the block.
+	// It verifies the receipt hash in the block header against the receipt hash of the fetched receipts
+	// to ensure that the execution engine did not fail to return any receipts.
+	FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (types.Receipts, error)
+}
+
+type CachingReceiptsProvider struct {
+	inner ReceiptsProvider
+	cache *caching.LRUCache[common.Hash, types.Receipts]
+}
+
+func NewCachingReceiptsProvider(inner ReceiptsProvider, m caching.Metrics, cacheSize int) *CachingReceiptsProvider {
+	return &CachingReceiptsProvider{
+		inner: inner,
+		cache: caching.NewLRUCache[common.Hash, types.Receipts](m, "receipts", cacheSize),
+	}
+}
+
+func (p *CachingReceiptsProvider) FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (types.Receipts, error) {
+	if r, ok := p.cache.Get(block.Hash); ok {
+		return r, nil
+	}
+
+	r, err := p.inner.FetchReceipts(ctx, block, txHashes)
+	if err != nil {
+		return nil, err
+	}
+
+	p.cache.Add(block.Hash, r)
+	return r, nil
+}
+
+// CachedReceipts provides direct access to the underlying receipts cache.
+// It returns the receipts and true for a cache hit, and nil, false otherwise.
+func (p *CachingReceiptsProvider) CachedReceipts(blockHash common.Hash) (types.Receipts, bool) {
+	return p.cache.Get(blockHash)
+}
+
+func newRPCRecProviderFromConfig(client client.RPC, log log.Logger, metrics caching.Metrics, config *EthClientConfig) *CachingReceiptsProvider {
+	recCfg := RPCReceiptsConfig{
+		MaxBatchSize:        config.MaxRequestsPerBatch,
+		ProviderKind:        config.RPCProviderKind,
+		MethodResetDuration: config.MethodResetDuration,
+	}
+	return NewCachingRPCReceiptsProvider(client, log, recCfg, metrics, config.ReceiptsCacheSize)
+}
+
+type rpcClient interface {
+	CallContext(ctx context.Context, result any, method string, args ...any) error
+	BatchCallContext(ctx context.Context, b []rpc.BatchElem) error
+}
+
+type BasicRPCReceiptsFetcher struct {
+	client       rpcClient
+	maxBatchSize int
+}
+
+func (f *BasicRPCReceiptsFetcher) FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (types.Receipts, error) {
+	fetcher := batching.NewIterativeBatchCall[common.Hash, *types.Receipt](
+		txHashes,
+		makeReceiptRequest,
+		f.client.BatchCallContext,
+		f.client.CallContext,
+		f.maxBatchSize,
+	)
+	// Fetch all receipts
+	for {
+		if err := fetcher.Fetch(ctx); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+	}
+	return fetcher.Result()
+}
```

Comment:
Rate-limit errors will stop this batch call, and then when restarting it, it will hit the rate limit errors again, and then never is able to finish fetching of the receipts data.

Reply:
I added back the preservation of batch calls in 52673275d7ea13289c1978b4aa42d8133a909afa. I also added better synchronization of fetching jobs to avoid the creation of duplicate fetching jobs. In the old implementation, if there were two racing concurrent accesses to the fetching jobs cache, there might have been created two concurrent fetching jobs for the same block hash.